File: typescript-playwright/src/api/SubcategoryApi.ts
import type { APIRequestContext, APIResponse } from "@playwright/test";
import { API_URL } from "../utils/constants";
import { mergeHeaders } from "../utils/headers";

export class SubcategoryApi {
  private readonly request: APIRequestContext;
  private readonly authHeaders: Record<string, string> = {};

  constructor(
    request: APIRequestContext,
    authHeaders: Record<string, string> = {},
  ) {
    this.request = request;
    this.authHeaders = authHeaders;
  }

  private get headers(): Record<string, string> {
    return mergeHeaders(this.authHeaders);
  }

  async getByCategoryId(categoryId: number): Promise<APIResponse> {
    return this.request.get(`${API_URL}/api/subcategories`, {
      params: { categoryId },
    });
  }

  async create(categoryId: number, name: string): Promise<APIResponse> {
    return this.request.post(`${API_URL}/api/subcategories`, {
      data: { categoryId, name },
      headers: this.headers,
    });
  }

  async deleteSubcategory(
    subcategoryId: number,
    headers?: Record<string, string>,
  ): Promise<APIResponse> {
    return this.request.delete(
      `${API_URL}/api/subcategories/${subcategoryId}`,
      {
        headers: headers ?? this.headers,
      },
    );
  }
}

File: python-pytest/src/api/subcategory_api.py
from __future__ import annotations

import requests

from src.utils.constants import API_URL
from src.utils.headers import merge_headers


class SubcategoryApi:
    def __init__(
        self, session: requests.Session, auth_headers: dict[str, str] | None = None
    ) -> None:
        self._session = session
        self._auth_headers = auth_headers or {}

    @property
    def _headers(self) -> dict[str, str]:
        return merge_headers(self._auth_headers)

    def get_by_category_id(self, category_id: int) -> requests.Response:
        return self._session.get(
            f"{API_URL}/api/subcategories",
            params={"categoryId": category_id},
            timeout=10,
        )

    def create(self, category_id: int, name: str) -> requests.Response:
        return self._session.post(
            f"{API_URL}/api/subcategories",
            json={"categoryId": category_id, "name": name},
            headers=self._headers,
            timeout=10,
        )

    def delete_subcategory(
        self, subcategory_id: int, headers: dict[str, str] | None = None
    ) -> requests.Response:
        return self._session.delete(
            f"{API_URL}/api/subcategories/{subcategory_id}",
            headers=headers or self._headers,
            timeout=10,
        )
